### PR TITLE
S0ix entry fix

### PR DESCRIPTION
**Pull request by keyonjie**
This is the replacement of @mrajwa 's https://github.com/thesofproject/linux/pull/2184. 

It is a simpler version, and aligned with SST driver.

### CODE REVIEW

File: sound/soc/sof/intel/hda-dsp.c [hunk 1]
```diff
@@ -696,12 +696,35 @@ int hda_dsp_resume(struct snd_sof_dev *sdev)
 		.state = SOF_DSP_PM_D0,
```
Comment (kv2019i) [inline sound/soc/sof/intel/hda-dsp.c:698]: Same comment I made to #2184 :
@mrajwa @keyonjie "I'd have opted for a shorter version and not go into details of PMC disqualifications, but this is good enough and ok for me. If you do another respin of the series, maybe explain what PMC means... (Power Management Controller) it's not a standard term in Linux kernel." But yeah, I gave approval once already, and I'll give again.

File: sound/soc/sof/intel/hda-dsp.c [hunk 2]
```diff
@@ -23,7 +23,7 @@
 #include "hda.h"
```
Comment (kv2019i) [inline sound/soc/sof/intel/hda-dsp.c]: @keyonjie @mrajwa Patch is good and I see a length discussion in https://github.com/thesofproject/linux/pull/2184/commits/765492ec2a28f2a5735edcf7c4955a0be1df7d1c . But, but, now the commit message would still need some polish before we send this out. It seems the commit ("ASoC: SOF: Intel: hda: Allow trace DMA in S0 when DSP is in D0I3 for debug") was incorrect. If CONFIG_SND_SOC_SOF_DEBUG is disabled, this patch now enables debugging by default. I.e. by enabling DEBUG option, the only thing you can do more, is to _disable_ traces when system stays in S0. This does not seem right... should we rather drop the option? @ranj063 ?

File: sound/soc/sof/intel/hda-dsp.c (hunk 1 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-dsp.c:698]: ack @kv2019i let me simply it.

File: sound/soc/sof/intel/hda-dsp.c (hunk 2 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-dsp.c]: @kv2019i good point about the condition about if CONFIG_SND_SOC_SOF_DEBUG is enabled or not. So, the condition to enable DMA Trace is SND_SOC_SOF_DEBUG_ENABLE_FIRMWARE_TRACE is selected ([1]sdev->dtrace_is_supported && [2]hda_enable_trace_D0I3_S0 && [3]suspend_target == S0). The code should be changed to be this:

```
		if (!sdev->dtrace_is_supported ||
		    !hda_enable_trace_D0I3_S0 ||
		    sdev->system_suspend_target != SOF_SUSPEND_NONE)
			flags = HDA_PM_NO_DMA_TRACE;
```

About the commit message, "ASoC: SOF: Intel: hda: Allow trace DMA in *S0* (condition [3] above) when *DSP is in D0I3* (condition [2] above, by default) for *debug* (condition [1] above)".

File: sound/soc/sof/intel/hda-dsp.c [hunk 3]
```diff
@@ -412,7 +412,7 @@ static int hda_dsp_set_D0_state(struct snd_sof_dev *sdev,
 		 * But it can be kept enabled when the DSP enters D0I3 while the
 		 * system is in S0 for debug.
 		 */
-		if (hda_enable_trace_D0I3_S0 &&
+		if (!hda_enable_trace_D0I3_S0 ||
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dsp.c]: The comment on line 411 says 'trace DMA is disabled by default'.
but you enable it by default in the kernel parameter.
One of the two is wrong then?

Comment (keyonjie) [reply sound/soc/sof/intel/hda-dsp.c]: Thanks, the line 411 actually should be "Trace DMA need to be disabled when the DSP enters D0I3 for S0Ix suspend".

Comment (plbossart) [reply sound/soc/sof/intel/hda-dsp.c]: then the value also needs to be 'false'?

File: sound/soc/sof/intel/hda-dsp.c [hunk 4]
```diff
@@ -23,7 +23,7 @@
 #include "hda.h"
 #include "hda-ipc.h"
 
-static bool hda_enable_trace_D0I3_S0;
+static bool hda_enable_trace_D0I3_S0 = true;
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dsp.c]: should be default false to match the code, this needs to be set as true for debug cases where you want to enable the trace in S0/D0i3.

File: sound/soc/sof/intel/hda-dsp.c (hunk 3 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-dsp.c]: you mean the value of hda_enable_trace_D0I3_S0? No, it denotes the value for S0 (not for S0Ix), we want to make DMA trace enabled by default on S0 for KWD (D0I3), when DMA trace is available, just same with other normal S0 scenarios.

Comment (plbossart) [reply sound/soc/sof/intel/hda-dsp.c]: @ranj063 did we really intend to have traces by default in S0/D0i3?

Comment (ranj063) [reply sound/soc/sof/intel/hda-dsp.c]: @plbossart no I dont think so. It is a debug feature and should only be enabled if requested.

Comment (keyonjie) [reply sound/soc/sof/intel/hda-dsp.c]: @ranj063 @plbossart I would ask the question in this way:

do we intend to have traces by default in S0/D0i3 when dtrace_is_supported is true (CONFIG_SND_SOC_SOF_DEBUG_ENABLE_FIRMWARE_TRACE=y, dma trace is enabled for normal S0)?

Or actually we should discard this "hda_enable_trace_D0I3_S0" item so no need for checking it (same with hda_enable_trace_D0I3_S0==true by default)?

File: sound/soc/sof/intel/hda-dsp.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/intel/hda-dsp.c:698]: @keyonjie Sorry, nitpicking again, but I complained earlier to @mrajwa that when this patch applies to all platforms (as it does), we should have some text in commit description on how widespread this problem is (to e.g. evalute whether the pach should be backported to kernels for old hw). The driver has worked e.g. on CML, so why are we changing the CML flow as well now. Otoh it is clear now driver is expected to shut down the link, so the patch is fixing the flow (even it has worked on some older platforms before due to firmware handling the missing steps driver should have done). If you respin, maybe put:

``
This patch fixes the suspend & resume procedure to allow entry into the
low power states with some streams being active as a wake source - wake on
voice is a perfect example. The current implementation does not stop 
the CORB/RIRB DMA and does not power down the HDA links. With firmware
help,  the platform has been able to still enter s0ix state on older platforms, but
the sequence is still incorrect and the additional driver actions are needed to ensure 
correct s0ix behaviour.
```

File: sound/soc/sof/intel/hda-dsp.c (hunk 4 above)
Comment (kv2019i) [reply sound/soc/sof/intel/hda-dsp.c]: @plbossart @keyonjie @ranj063  Hmm, I'm also not really clear whether we really have a need for this rich interface to control via kernel params. I'd imagine for most case, simple check of:
```
	if (!sdev->dtrace_is_supported ||
	    sdev->system_suspend_target != SOF_SUSPEND_NONE)
			flags = HDA_PM_NO_DMA_TRACE;
```

... would cover most needs without any extra kernel params.

But alas, my mileage debugging d0ix issues pales in comparison to you folks, so I'll take your word on this. I do think the default should be "false".  D0ix with DMA active for traces is a special state, so should not be the default.

**Comment by keyonjie** [comment]
Thanks @kv2019i @plbossart I will respin them will leave the default false for hda_enable_trace_D0I3_S0 and adding commit message as @kv2019i suggested.

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by fredoh9** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
> SOFCI TEST

Wondering why nothing happens when I do this... Have I lost my superpowers? Or am I on a 'denylist'?

**Comment by plbossart** [comment]
known CI issues unrelated to this PR, merging

**Pull request merged**